src/main/java/com/mitchej123/hodgepodge/SimulationDistanceHelper.java

```diff
@@ -0,0 +1,332 @@
+
+    /**
+     * Cache of forced chunks. Profiler says that is faster.
+     */
+    private final LongOpenHashSet forcedChunksMap = new LongOpenHashSet();
+
+    /**
+     * Cache to quickly check if a chunk should be processed or not. Profiler says this is faster.
+     */
+    private final Long2BooleanOpenHashMap cacheShouldProcessTick = new Long2BooleanOpenHashMap();
+
+    /**
+     * Old player positions to see if they moved.
+     */
+    private final Map<EntityPlayer, ChunkCoordIntPair> playerPosOldMap = new HashMap<>();
+
+    /**
+     * All ticks that are currently considered.
+     */
+    private final TreeSet<NextTickListEntry> pendingTickCandidates = new TreeSet<>();
+
+    /**
+     * Map of ticks to chunk, duplicating pendingTickListEntriesTreeSet for quicker lookup.
+     */
+    private final Long2ObjectMap<HashSet<NextTickListEntry>> chunkTickMap = new Long2ObjectOpenHashMap<>();
+
+    /**
+     * TreeSet pendingTickListEntriesTreeSet from WorldServer
+     */
+    private TreeSet<NextTickListEntry> pendingTickListEntriesTreeSet;
+    /**
+     * HashSet pendingTickListEntriesHashSet from WorldServer
+     */
+    private Set<NextTickListEntry> pendingTickListEntriesHashSet;
+    /**
+     * Check if chunk exists in WorldServer
+     */
+    private BiPredicate<Integer, Integer> chunkExists;
+
+    /**
+     * Simulation distance from last tick, to detect if it has changed.
+     */
+    private int simulationDistanceOld;
+
+    public void preventChunkSimulation(long packedChunkPos, boolean prevent) {
+        synchronized (noTickChunksChanges) {
+            noTickChunksChanges.add(LongBooleanPair.of(packedChunkPos, prevent));
+        }
+    }
+
+    private boolean closeToPlayer(int x, int z) {
+        int simulationDistance = getSimulationDistance();
+        for (EntityPlayer player : world.playerEntities) {
+            if (player.getEntityWorld() != world) {
+                continue;
+            }
+            int playerX = (int) player.posX >> 4;
+            int playerZ = (int) player.posZ >> 4;
+            if (Math.abs(playerX - x) <= simulationDistance && Math.abs(playerZ - z) <= simulationDistance) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a chunk should get processed
+     */
+    public boolean shouldProcessTick(int x, int z) {
+        long key = ChunkCoordIntPair.chunkXZ2Int(x, z);
+        return cacheShouldProcessTick.computeIfAbsent(key, dummy -> {
+            if (closeToPlayer(x, z)) {
+                return true;
+            }
+            if (noTickChunks.containsKey(key)) {
+                return false;
+            }
+            return forcedChunksMap.contains(key);
+        });
+    }
+
+    private LongOpenHashSet getPlayerChunksForPos(ChunkCoordIntPair pos) {
+        int simulationDistance = getSimulationDistance();
+        LongOpenHashSet chunks = new LongOpenHashSet();
```

Comment:
Pass a capacity to the constructor to avoid reallocations (2 * simulationDistance + 1 squared)

Reply:
Thanks, pushed an update